xds/src/main/java/io/grpc/xds/XdsClientImpl.java

```diff
@@ -187,6 +206,65 @@ void watchConfigData(String hostName, int port, ConfigWatcher watcher) {
     adsStream.sendXdsRequest(ADS_TYPE_URL_LDS, ImmutableList.of(ldsResourceName));
   }
 
+  @Override
+  void watchClusterData(String clusterName, ClusterWatcher watcher) {
+    checkNotNull(watcher, "watcher");
+    boolean needRequest = false;
+    if (!clusterWatchers.containsKey(clusterName)) {
+      logger.log(Level.FINE, "Start watching cluster {0}", clusterName);
+      needRequest = true;
+      clusterWatchers.put(clusterName, new HashSet<ClusterWatcher>());
+    }
+    Set<ClusterWatcher> watchers = clusterWatchers.get(clusterName);
+    if (watchers.contains(watcher)) {
+      logger.log(Level.WARNING, "Watcher {0} already registered", watcher);
+    }
+    watchers.add(watcher);
+    // If local cache contains cluster information to be watched, notify the watcher immediately.
+    if (clusterNamesToClusterUpdates.containsKey(clusterName)) {
```

Comment:
Few comments on the logic here:

- if the watcher is already present (line 219 returns true) can you just return?
- how about replacing lines 219-222 with
```
    if (!watchers.add(watcher)) {
      logger.log(Level.WARNING, "Watcher {0} already registered", watcher);
    }
```
- if the watcher is already registered, do you still need to send an immediate update (this would be a duplicate update). Depends on the semantics of watching.

Reply:
First of all, users of this class should know what `XdsClient` is doing and how should the API be used, especially for such an obvious API (for what reason would a user add the same watcher for the same cluster twice?). They should be aware by themselves that they are probably doing something not correct if adding the same watcher for the same cluster more than once. The API provider does not necessarily give them correct outcome when they use it incorrectly.

Here, yes, I agree changing it to return directly looks better, as it's easy and cheap to still give the user correct outcome even if they use it incorrectly.

Reply:
Changed. Adding the same watcher for the same cluster more than once is a no-op (just log with warning), canceling non-existing watcher is also a no-op (just log with warning).

Reply:
> ... for such an obvious API (for what reason would a user add the same watcher for the same cluster twice?). They should be aware by themselves that they are probably doing something not correct if adding the same watcher for the same cluster more than once. 
> ...
> Here, yes, I agree changing it to return directly looks better, as it's easy and cheap to still give the user correct outcome even if they use it incorrectly.

It's up to you to define the semantics of your API. If you don't want or expect your callers to pass the same watcher more than once I think it's better to make it an error for the subsequent calls so callers clean up their code (instead of silently returning like you are doing). Also see my comment below for remove.